olive/engine/packaging/packaging_generator.py

```diff
@@ -32,164 +37,249 @@
-                    packaging_config.export_in_mlflow_format,
-                )
-        _package_onnxruntime_packages(tempdir, next(iter(pf_footprints.values())))
-        shutil.make_archive(packaging_config.name, "zip", tempdir)
-        package_file = f"{packaging_config.name}.zip"
-        shutil.move(package_file, output_dir / package_file)
-
-
-def _package_models_rank(tempdir, footprints: Dict["AcceleratorSpec", "Footprint"]):
-    metrics_dict = next(iter(footprints.values())).objective_dict
-    sorted_nodes = sorted(
-        itertools.chain.from_iterable(f.nodes.values() for f in footprints.values()),
-        key=lambda x: tuple(
-            x.metrics.value[metric].value if x.metrics.cmp_direction[metric] == 1 else -x.metrics.value[metric].value
-            for metric in metrics_dict
-        ),
-        reverse=True,
+            footprint = footprints[accelerator_spec]
+            if pf_footprint.nodes and footprint.nodes:
+                model_rank = 1
+                input_node = footprint.get_input_node()
+                for model_id, node in pf_footprint.nodes.items():
+                    model_name = f"{output_name}_{accelerator_spec}_{model_rank}"
+                    if packaging_type == PackagingType.Zipfile:
+                        model_dir = (
+                            tempdir / "CandidateModels" / str(accelerator_spec) / f"BestCandidateModel_{model_rank}"
+                        )
+                    else:
+                        model_dir = tempdir / model_name
+
+                    model_dir.mkdir(parents=True, exist_ok=True)
+
+                    # Copy inference config
+                    inference_config_path = model_dir / "inference_config.json"
+                    inference_config = pf_footprint.get_model_inference_config(model_id) or {}
+
+                    _copy_inference_config(inference_config_path, inference_config)
+                    _copy_configurations(model_dir, footprint, model_id)
+                    _copy_metrics(model_dir, input_node, node)
+                    _save_model(pf_footprint, model_id, model_dir, inference_config, export_in_mlflow_format)
+
+                    model_info_list = []
+                    model_info = _get_model_info(node, model_rank)
+                    model_info_list.append(model_info)
+                    _copy_model_info(model_dir, model_info)
+
+                    if packaging_type == PackagingType.AzureMLModels:
+                        _upload_to_azureml_models(azureml_client_config, model_dir, model_name, config)
+                    elif packaging_type == PackagingType.AzureMLData:
+                        _upload_to_azureml_data(azureml_client_config, model_dir, model_name, config)
+
+                model_rank += 1
+
+        if packaging_type == PackagingType.Zipfile:
+            _copy_models_rank(tempdir, model_info_list)
+            _package_zipfile_model(output_dir, output_name, tempdir)
+
+
+def _upload_to_azureml_models(
```

Comment:
seems the name is upload multiple models but the function only support one model?

Reply:
This "models" mainly refers to the tag name in Azure ML workspace. It is called "Models" in AML workspace.

Reply:
could we add some comments for the naming, it is a little bit confused since Olive also has model concept.

Reply:
Yeah I can add a comment here.